### PR TITLE
Document, test, and harden the Falcon math module

### LINKED ISSUE #3575: crypto: Polynomial::karatsuba panics on empty and unequal-length operands
### Packages versions

miden-crypto 0.30.0 (next, commit d69d25f4)

### Bug description

`Polynomial::karatsuba` panics on inputs that `Mul` handles correctly. Both
compute the same product, but `Mul::mul` guards with `is_zero()` while
`karatsuba` passes its coefficients straight to `vector_karatsuba`.

Two cases, both reachable through the public API
(`miden_crypto::dsa::falcon512_poseidon2::Polynomial` is re-exported at
falcon512_poseidon2/mod.rs:50, and `karatsuba` is `pub fn`):

1. Empty operands. `vector_karatsuba` computes
   `left.len() + right.len() - 1`, which underflows at 0 + 0.
   polynomial.rs:480 — `attempt to subtract with overflow` in debug,
   capacity overflow in release.

2. Unequal lengths. `vector_karatsuba` splits both slices at
   `n_over_2` derived from `left.len()` only, so a shorter right operand
   is indexed out of bounds. polynomial.rs:491 —
   `range end index 8 out of range for slice of length 2`.
   `Mul` handles unequal lengths fine via its nested loops.

### How to reproduce

From a downstream crate depending on miden-crypto:

    let a = Polynomial::<i64>::new(vec![]);
    let b = Polynomial::<i64>::new(vec![]);
    let _ = a.karatsuba(&b);          // panics
    let _ = a.clone() * b.clone();    // returns zero, no panic

    let a = Polynomial::<i64>::new(vec![1; 16]);
    let b = Polynomial::<i64>::new(vec![1; 2]);
    let _ = a.karatsuba(&b);          // panics

I confirmed all three locally against a probe crate: both karatsuba calls
panic, the `Mul` control passes.

### Expected behaviour

`karatsuba` should mirror `Mul`: return the zero polynomial for zero
operands, and either handle unequal lengths or reject them explicitly
rather than indexing out of bounds.

Not reachable from untrusted input — `SignaturePoly::read_from` decodes into
a fixed `[FalconFelt; N]`, and the internal callers (`ntru_solve`,
`babai_reduce`) always pass matched 512-coefficient polynomials. This is a
public-API robustness issue, in the same family as #3534.

I'd like to work on this if it can be assigned to me. The fix is adding the
`is_zero()` guard and a length check, plus regression tests for both cases.

### LINKED ISSUE #3607: bug: Polynomial::div never terminates when the leading-coefficient division truncates to zero
## Describe your changes

`Polynomial<F>`'s `Div` impl does long division, and the loop only makes progress when the quotient of the leading coefficients is non-zero. When `F` is an integer type and `lc(remainder) / lc(denominator)` truncates to `0`, the subtracted monomial is zero, the remainder stops changing, and the loop spins forever.

Reproduction — this never returns:

```rust
use miden_crypto::dsa::falcon512_poseidon2::math::Polynomial;

// (1 + 3x) / 2
let _ = Polynomial::new(vec![1i64, 3]) / Polynomial::new(vec![2i64]);
```

Stepping through it: the first iteration subtracts `2x` and leaves `1 + x`; from the second iteration on, `1 / 2` truncates to `0`, so the monomial is zero, `remainder` stays at `1 + x`, and `is_zero()` never fires.

Field types cannot hit this: the leading coefficient of a non-zero remainder is non-zero by definition, and dividing it by a non-zero element of a field is never zero, so the degree strictly decreases every iteration. But nothing on the generic impl says it is field-only, and `div_zero_by_nonzero_returns_zero` (added in #3573) exercises `Div` with `Polynomial<i64>`.

Related: #3573 fixed the zero-numerator panic in the same function.

### LINKED ISSUE #3621: Polynomial::galois_adjoint is documented as f(x^2) but computes f(-x)
`Polynomial::galois_adjoint` is documented as computing `f(x^2)`:

```rust
/// Computes the galois adjoint of the polynomial in the cyclotomic ring F[ X ] / < X^n + 1 >
/// , which corresponds to f(x^2).
pub fn galois_adjoint(&self) -> Self {
```

It computes `f(-x)`. The body keeps the even-power coefficients and negates the odd-power ones, which is the substitution `x -> -x`, since `f(-x) = sum c_i (-1)^i x^i`.

`f(x^2)` is what the neighbouring `lift_next_cyclotomic` does, by spreading the coefficients over the even powers of a ring of twice the size.

`ntru_solve` uses both and its local names agree with that reading:

```rust
let capital_f_prime_xsq = capital_f_prime.lift_next_cyclotomic();
let capital_g_prime_xsq = capital_g_prime.lift_next_cyclotomic();

let f_minx = f.galois_adjoint();
let g_minx = g.galois_adjoint();
```

So the code is correct and follows the Falcon spec (`F = F'(x^2) * g(-x)`, `G = G'(x^2) * f(-x)`); only the doc comment on `galois_adjoint` is wrong, and it happens to describe what `lift_next_cyclotomic` does.

Neither function has a unit test, so nothing pins the intended substitution either way.

### Suggested fix

Correct the comment on `galois_adjoint` to `f(-x)`, mention `f(x^2)` on `lift_next_cyclotomic`, and add a test for each pinning the substitution.

### PR DESCRIPTION
**Pull request by Al-Kindi-0**
Closes #3487.

**Breaking:** `Polynomial::karatsuba` now rejects unequal, empty, and unsupported recursive operand shapes with documented panics. Some unequal operands at the old base case previously happened to work.

Documents the Falcon `math` module and adds oracle-style test coverage throughout (module tests 19 -> 54): SamplerZ reference known-answer vectors, RCDT threshold boundaries, FFT table self-derivation and convolution checks, and NTRU equation checks.

Writing these surfaced a few edge-case defects, now fixed: `FalconFelt::new` canonicalization for negative multiples of q, `Polynomial::div` non-termination when a division step fails to cancel the leading term, unenforced `karatsuba`/FFT size preconditions, and several incorrect doc comments (e.g. `galois_adjoint` computes `f(-x)`, not `f(x^2)`). None of the fixes change behavior on live Falcon signing/keygen paths.

### CODE REVIEW

File: crates/crypto/src/dsa/falcon512_poseidon2/math/polynomial.rs [hunk 1]
```diff
@@ -377,7 +380,21 @@ impl<F: Mul<Output = F> + Sub<Output = F> + AddAssign + Zero + Div<Output = F> +
     Polynomial<F>
 {
     /// Multiply two polynomials using Karatsuba's divide-and-conquer algorithm.
+    ///
+    /// Both coefficient vectors must have the same nonzero length `n`, and `n` must stay even
+    /// under repeated halving until it reaches the recursion's base case (`n <= 8`); any power of
+    /// two qualifies, and Falcon only multiplies power-of-two lengths. Unsupported shapes panic:
+    /// an odd split truncates a cross term and overruns the output buffer.
     pub fn karatsuba(&self, other: &Self) -> Self {
+        assert_eq!(
```
Comment (huitseeker) [inline crates/crypto/src/dsa/falcon512_poseidon2/math/polynomial.rs:393]: `Polynomial::karatsuba` now panics for unequal operands that the old base case multiplied correctly, such as lengths 8 and 4. 

Since `Polynomial` is publicly re-exported, could this PR be marked as breaking, document the limitation, drop the claim that karatsuba matches `Mul` and include a `Panic` comment here?

File: crates/crypto/src/dsa/falcon512_poseidon2/math/fft.rs [hunk 2]
```diff
@@ -248,22 +249,30 @@ impl CyclotomicFourier for Complex64 {
...
+}
+
 impl FastFft for Polynomial<Complex64> {
     type Field = Complex64;
     fn fft_inplace(&mut self) {
         let n = self.coefficients.len();
-        debug_assert!(
-            (1..=512).contains(&n),
-            "unsupported: n = {n} not a power of 2 or larger than 512"
+        assert!(
```
Comment (huitseeker) [inline crates/crypto/src/dsa/falcon512_poseidon2/math/fft.rs]: nit: The two `FastFft` impls repeat their validation logic and duplicate panic tests. Could the generic `CyclotomicFourier` helpers validate against each supplied table length so direct callers and both impls share one contract?

File: crates/crypto/src/dsa/falcon512_poseidon2/math/field.rs [hunk 3]
```diff
@@ -5,29 +5,40 @@ use num::{One, Zero};
...
     pub const fn new(value: i16) -> Self {
-        let gtz_bool = value >= 0;
-        let gtz_int = gtz_bool as i16;
-        let gtz_sign = gtz_int - ((!gtz_bool) as i16);
-        let reduced = gtz_sign * (gtz_sign * value) % MODULUS;
-        let canonical_representative = (reduced + MODULUS * (1 - gtz_int)) as u32;
+        // Truncated remainder lands in (-q, q); adding q and reducing once more lands in [0, q)
+        // for every input, including negative multiples of q.
+        let reduced = value as i32 % MODULUS as i32;
+        let canonical_representative = ((reduced + MODULUS as i32) % MODULUS as i32) as u32;
```
Comment (huitseeker) [inline crates/crypto/src/dsa/falcon512_poseidon2/math/field.rs]: nit: `i16::rem_euclid` already returns the canonical representative in `[0, MODULUS)`. Could `new` use `FalconFelt(value.rem_euclid(MODULUS) as u32)` so the standard operation carries this invariant without the two remainder steps?

File: crates/crypto/src/dsa/falcon512_poseidon2/math/samplerz.rs [hunk 4]
```diff
@@ -127,7 +134,183 @@ pub(crate) fn sampler_z<R: Rng>(mu: f64, sigma: f64, sigma_min: f64, rng: &mut R
...
+    /// decide on the first byte, so without this a corruption confined to a deeper comparison
+    /// byte survives every other test in the crate (mutation-confirmed).
+    #[test]
+    fn ber_exp_decides_at_every_comparison_depth() {
+        // Parameters chosen so s = floor(x / ln 2) = 0 and every comparison byte of z is
+        // strictly interior; the assert below makes any platform float drift loud instead of
+        // silently weakening the test.
+        let (x, ccs) = (0.3f64, 0.7f64);
+        let z = (approx_exp(x, ccs) << 1) - 1;
+        let z_bytes: Vec<u8> = (0..8).map(|j| ((z >> (56 - 8 * j)) & 0xff) as u8).collect();
```
Comment (huitseeker) [inline crates/crypto/src/dsa/falcon512_poseidon2/math/samplerz.rs]: nit: `ReplayRng::new` can use `hex::decode(hex)`, and this byte loop can use `z.to_be_bytes().to_vec()`. Could we use those existing conversions to keep the KAT harness focused on sampler behavior?

File: crates/crypto/src/dsa/falcon512_poseidon2/math/polynomial.rs [hunk 5]
```diff
@@ -662,6 +710,99 @@ mod tests {
...
+
+    #[test]
+    fn karatsuba_matches_schoolbook_for_every_accepted_shape() {
+        // Powers of two (the live Falcon shapes) plus the accepted non-power-of-two lengths,
+        // which halve evenly to the base case and would otherwise have no output coverage.
+        for n in [2i64, 4, 8, 16, 32, 10, 20, 24] {
+            let f = Polynomial::new((0..n).map(|i| i * i - 7 * i + 3).collect());
+            let g = Polynomial::new((0..n).map(|i| 5 * i - 11).collect());
+            let schoolbook = f.clone() * g.clone();
+            assert_eq!(f.karatsuba(&g), schoolbook, "karatsuba disagrees at n = {n}");
```
Comment (huitseeker) [inline crates/crypto/src/dsa/falcon512_poseidon2/math/polynomial.rs:753]: nit: This checks one coefficient pattern for each accepted equal length. Since `proptest` is already a dev dependency, could we generate independent nonempty compatible operand lengths and assert `karatsuba == Mul`?

**Review by huitseeker — COMMENTED** [review]
Thanks! overall this makes sense, though we could use proptest to close testing gaps, and make admissible inputs and panic conditions clear.

File: crates/crypto/src/dsa/falcon512_poseidon2/math/polynomial.rs (hunk 1 above)
Comment (Al-Kindi-0) [reply crates/crypto/src/dsa/falcon512_poseidon2/math/polynomial.rs:393]: Addressed in 04934e744. The changelog and PR body now mark this as breaking, `karatsuba` has an explicit `# Panics` contract, and equality claims/tests are scoped to admissible equal-length operands.

File: crates/crypto/src/dsa/falcon512_poseidon2/math/fft.rs (hunk 2 above)
Comment (Al-Kindi-0) [reply crates/crypto/src/dsa/falcon512_poseidon2/math/fft.rs]: Addressed in 04934e744. Validation now lives in the generic `CyclotomicFourier` FFT/IFFT/split/merge helpers and uses the supplied table length. The duplicated implementation checks and type-specific panic tests are gone; a direct short-table regression covers the generic contract.

File: crates/crypto/src/dsa/falcon512_poseidon2/math/field.rs (hunk 3 above)
Comment (Al-Kindi-0) [reply crates/crypto/src/dsa/falcon512_poseidon2/math/field.rs]: Addressed in 04934e744: `FalconFelt::new` now uses `value.rem_euclid(MODULUS)` directly.

File: crates/crypto/src/dsa/falcon512_poseidon2/math/samplerz.rs (hunk 4 above)
Comment (Al-Kindi-0) [reply crates/crypto/src/dsa/falcon512_poseidon2/math/samplerz.rs]: Addressed in 04934e744. The replay harness now uses `hex::decode`, and the comparison bytes use `z.to_be_bytes().to_vec()`.

File: crates/crypto/src/dsa/falcon512_poseidon2/math/polynomial.rs (hunk 5 above)
Comment (Al-Kindi-0) [reply crates/crypto/src/dsa/falcon512_poseidon2/math/polynomial.rs:753]: Addressed in 04934e744. Added a proptest that generates independent coefficient vectors at nonempty admissible lengths up to 512 and checks `karatsuba == Mul`. A recursive-result mutation makes the focused property fail and shrink a counterexample.

File: CHANGELOG.md [hunk 6]
```diff
@@ -8,6 +8,8 @@
 #### Changes
 
 - [BREAKING] `UniqueNodes` entries are now keyed by tree position, and missing nodes mean canonical empty subtree roots. The `NodeValue` enum was removed ([#3620](https://github.com/0xMiden/miden-vm/pull/3620)).
+- Documented, tested, and hardened the Falcon `math` module: fixed `FalconFelt::new` canonicalization and `Polynomial::div` non-termination edge cases, enforced `karatsuba` and FFT size preconditions, and added oracle-style tests (SamplerZ reference vectors, RCDT boundaries, FFT self-derivation and convolution checks, NTRU equation checks) ([#3629](https://github.com/0xMiden/miden-vm/pull/3629)).
```
Comment (huitseeker) [inline CHANGELOG.md]: nit: This duplicates the Falcon entry immediately above, so the changelog describes PR #3629 twice. Could we keep only the breaking entry?

**Review by huitseeker — COMMENTED** [review]
Apart from the Changelog, lgtm!

**Review by huitseeker — APPROVED** [review]
TYSVM!
